### PR TITLE
Create SOP for knative_up != 1

### PR DESCRIPTION
**Pull request by aliok**
Create a simple SOP for the case where `knative_up` metric is not `1`.

Users/services can create an alert on their own, using this metric and can use this document as a SOP for that alert.

cc @ReToCode 
cc @lberk 
cc @skonto 

@jewzaam is this a good SOP? Could please take a look?

@danielezonca is this usable for your project?

### CODE REVIEW

File: docs/sops/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# OpenShift Serverless SOPS
```
Comment (matzew) [inline docs/sops/README.md:1]: SOPS: *S*tandard *O*perating *P*rocedure*s* ?

File: docs/sops/serverless-controller-knative-up.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+1. Check to see if all the pods are running in `openshift-serverless` namespace:
+
+```bash
+$ oc -n openshift-serverless get pods
+```
+
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
+
+```bash
+# Get Pod names as we don't have a common label
```
Comment (matzew) [inline docs/sops/serverless-controller-knative-up.md:36]: should we introduce such a label? (wondering about this being scriptable)

File: docs/sops/serverless-controller-knative-up.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+This failure should only occur in extremely catastrophic scenarios as all components are deployed in HA configuration.
+
+Possible causes can be:
+- Network issues in the cluster
+- Resource starvation on the cluster
+- OpenShift Serverless upgrade issues
+
+You should check both OpenShift Serverless operator and Knative Serving control plane status.
+
+## Steps
```
Comment (danielezonca) [inline docs/sops/serverless-controller-knative-up.md:25]: Let's add a "Prerequisite" sections, in this case I expect it is enough to have the cluster and a `oc` client logged in but it is a good template to follow

File: docs/sops/serverless-controller-knative-up.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+## Impact
+
+Any users or services using Knative Serving will be unable to get their Knative Services reconciled.
+This means, any new Knative Services might not be created, and any existing Knative Services might not be updated. Also, it might also mean that the user workload is not being served.
+
+## Summary
+
+OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the metrics exposed.
+
+The metric `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`.
```
Comment (danielezonca) [inline docs/sops/serverless-controller-knative-up.md:14]: I would personally create a simple list of alerts (in this repo?) that users can create so that here we can refer it also by name

File: docs/sops/serverless-controller-knative-up.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+You should check both OpenShift Serverless operator and Knative Serving control plane status.
+
+## Steps
+
+1. Check to see if all the pods are running in `openshift-serverless` namespace:
+
+```bash
+$ oc -n openshift-serverless get pods
+```
```
Comment (danielezonca) [inline docs/sops/serverless-controller-knative-up.md:31]: I would make this explicit: if the goal is to get all pods that are not running make the command already filter out only them like `--field-selector status.phase!=Running`

The reason of this comment is that this procedure should be as much as possible simple, for example if the command is going to return a lot of pods you might risk to don't see one of them not running

File: docs/sops/serverless-controller-knative-up.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+$ oc -n openshift-serverless rollout restart deployments/knative-openshift-ingress
+$ oc -n openshift-serverless rollout restart deployments/knative-operator-webhook
+```
+
+This should result in new pods getting deployed, attempt step (1) again and see if the pods achieve running state.
+
+5. Check to see if all the pods are running in `knative-serving` namespace:
+
+```bash
+$ oc -n knative-serving get pods -l app.kubernetes.io/name=knative-serving
```
Comment (danielezonca) [inline docs/sops/serverless-controller-knative-up.md:67]: Same as above

**Review by danielezonca — COMMENTED** [review]
Thanks @aliok for the PR, overall I think it is a good example of SOP, I have some comments but overall it looks good to me :)

**Comment by matzew** [comment]
Is this generally something we want to "proper" document?  (e.g. on some "openshift docs" format)?

Or are we stashing these documents here and eventually move them else?

File: docs/sops/serverless-controller-knative-up.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+## Steps
+
+1. Check to see if all the pods are running in `openshift-serverless` namespace:
+
+```bash
+$ oc -n openshift-serverless get pods
+```
+
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
```
Comment (ReToCode) [inline docs/sops/serverless-controller-knative-up.md:33]: I'm not sure aware of the target audience for a SOP, but I'd assume that anyone operating the cluster already knows how to do to this? Would it be sufficient to just list which pods to take a look at, similar to 
https://knative.dev/docs/install/troubleshooting/

File: docs/sops/serverless-controller-knative-up.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
+# OpenShift Serverless Knative Serving Control Plane Readiness
```
Comment (ReToCode) [inline docs/sops/serverless-controller-knative-up.md:1]: This seems like it is targeting only the operator pods. So that is not really "Serving", right? 
For Serving we'd also need to look at all pods in `knative-serving`, `knative-serving-ingress` and (when istio is used) `istio-system` (or a custom namespace where istio is deployed).

File: docs/sops/serverless-controller-knative-up.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+7. Redeploy Knative Serving controllers by restarting the deployments:
+
+```bash
+$ oc -n knative-serving rollout restart deployments -l app.kubernetes.io/name=knative-serving
+```
+
+This should result in new pods getting deployed, attempt step (5) again and see if the pods achieve running state.
+
+8. If the problem persists, capture the logs and escalate to OpenShift Serverless engineering team.
```
[the remark this answers is not in the document]
Comment (ReToCode) [reply docs/sops/serverless-controller-knative-up.md:91]: Yes, although a non-knative specific must-gather would help as well.

**Review by ReToCode — COMMENTED** [review]
I like it. I'm also wondering about https://github.com/openshift-knative/serverless-operator/pull/2272#issuecomment-1727019166.

File: docs/sops/serverless-controller-knative-up.md (hunk 7 above)
Comment (matzew) [reply docs/sops/serverless-controller-knative-up.md:33]: I like that idea from upstream, instead of doing too much of an introduction, which can be assumed as known (I guess)

File: docs/sops/serverless-controller-knative-up.md (hunk 2 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:36]: Created https://issues.redhat.com/browse/SRVCOM-2703

File: docs/sops/README.md (hunk 1 above)
Comment (aliok) [reply docs/sops/README.md:1]: updated

File: docs/sops/serverless-controller-knative-up.md (hunk 3 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:25]: I copy pasted the ModelServing SOP in the private SRE repo and used it as the template, but let me add a pre-reqs section. It is a good idea.

File: docs/sops/serverless-controller-knative-up.md (hunk 8 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:1]: Down in the file, there are instructions to check if pods in `knative-serving` is running.

But, let me add the same thing for other namespaces.

Thanks

File: docs/sops/serverless-controller-knative-up.md (hunk 7 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:33]: I've followed the SOPs in SRE repository in managed tenants. I would prefer keeping this format. It is them who is going to use the SOPs.

File: docs/sops/serverless-controller-knative-up.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:95]: The idea to escalate to the engineering team with the _logs_. 
Then if eng team asks for a must-gather after checking the collected logs, SREs can do it.

So, IMO, let's not add it here.

File: docs/sops/serverless-controller-knative-up.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+# Check pod logs
+$ for element in "${SERVERLESS_OPERATOR_POD_NAMES[@]}"
+$ do
+$     oc -n openshift-serverless logs "$element" --prefix=true
+$ done
+
+
+# Check events 
+$ oc -n openshift-serverless get events
```
[the remark this answers is not in the document]
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:52]: I do not have that information right now.
Once we have it under operation, we can add some known events gradually.

**Comment by aliok** [comment]
@matzew  @ReToCode 

> Is this generally something we want to "proper" document? (e.g. on some "openshift docs" format)?
> Or are we stashing these documents here and eventually move them else?

The idea is that other teams would need this SOP in the future. I have tried pushing the same SOP in the SRE repository hoping this would help with collaboration, but it didn't.

Now, I am preparing a single SOP in advance, which will be needed. And once it is needed, it can be consumed pretty easily. We will need more SOPs, but that's something I need to work on with other teams.

> Or are we stashing these documents here and eventually move them else?
So, basically stashing these documents here and later move them somewhere else.

File: docs/sops/serverless-controller-knative-up.md (hunk 8 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:1]: Added some text for Ingress now.

Thanks @ReToCode for the suggestion.

File: docs/sops/serverless-controller-knative-up.md (hunk 5 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:58]: I would prefer showing all pods.

Reason:
- Later on, there's an instruction to collect logs from pods. The `oc logs` command doesn't support something like "only logs of failing/unready/crashing pods" (it doesn't make sense to only collect the logs of those anyway. IMO, logs of the whole system should be collected)
- So, the person will collect the logs will see all logs, but they will see new pods that they haven't seen before.

We don't have too many pods anyway.

Also, the SOP I've used as a basis checks all logs, and not only the failing ones.

If one day and SRE comes and says we need to go this route, I can do it. I feel it is a bit of premature optimization at this point.

File: docs/sops/serverless-controller-knative-up.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+```bash
+$ oc -n openshift-serverless get pods
+```
+
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
+
+```bash
+# Get Pod names as we don't have a common label
+$ SERVERLESS_OPERATOR_POD_NAMES_STR=$(oc -n openshift-serverless get pods --template '{{range .items}}{{.metadata.name}}{{" "}}{{end}}')
```
[the remark this answers is not in the document]
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:64]: Same as https://github.com/openshift-knative/serverless-operator/pull/2272#discussion_r1332718317

**Comment by aliok** [comment]
@ReToCode @danielezonca 
Thanks very much for your reviews. I really appreciate them.

Can you have a new look please? I've addressed all of your comments (made code changes OR replied with my reasoning)

File: docs/sops/serverless-controller-knative-up.md (hunk 4 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:14]: Good idea @danielezonca !

I've created a ticket for that now: https://issues.redhat.com/browse/SRVCOM-2707

I think that's a separate issue/work.

**Review by ReToCode — COMMENTED** [review]
A thing I noted when checking the markdown view, adding `$` is not ideal if you want to copy the commands. Especially if they are multi-line. I think dropping that would help with UX here.

Other than that, LGTM.

**Comment by aliok** [comment]
>  I think dropping that would help with UX here.

I've dropped the `$` prefix for multiline scripts now.

**Comment by skonto** [comment]
Not sure if anyone mentioned this but `knative_up` is a gauge and users should be aware of what prom query they should use: https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/.

File: docs/sops/serverless-controller-knative-up.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+## Impact
+
+Any users or services using Knative Serving will be unable to get their Knative Services reconciled.
+This means, any new Knative Services might not be created, and any existing Knative Services might not be updated. Also, it might also mean that the user workload is not being served.
+
+## Summary
+
+OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the metrics exposed.
+
+The metric `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`.
```
Comment (skonto) [inline docs/sops/serverless-controller-knative-up.md:14]: Actually for gauges it is better to measure it over some period of time eg. https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/.

Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:14]: Thanks for the great input @skonto !

`avg_over_time(up{job="jobname"}[5m]) < 0.9` we should use this kind of structure in https://issues.redhat.com/browse/SRVCOM-2704

Comment (aliok) [inline docs/sops/serverless-controller-knative-up.md:14]:
```suggestion
OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the gauges and metrics exposed.

The gauge `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`.  You should create an alert as described in https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/ for using this gauge and measure it over some period of time.
```

Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md]: Also updated the text here to reflect that `knative_up` is a gauge and consumers should measure it over period of time. However, we don't provide alerts in this PR. That's part of SRVCOM-2704.

File: docs/sops/serverless-controller-knative-up.md [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+```yaml
+  contour:
+    enabled: false
+  istio:
+    enabled: true
+  kourier:
+    enabled: false
+```
+
+In this case, Istio is used as the Knative Serving Ingress.
```
Comment (skonto) [inline docs/sops/serverless-controller-knative-up.md:50]: Probably we should mention that we use ServiceMesh and people should also check their setup there for issues.

File: docs/sops/serverless-controller-knative-up.md [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+```bash
+$ oc -n knative-serving-ingress rollout restart deployments -l app.kubernetes.io/component=net-kourier
+```
+
+This should result in new pods getting deployed, attempt step (8) again and see if the pods achieve running state.
+
+11. If Istio is used as the Knative Serving Ingress, check Istio status. You may find the SOPs for Istio by contacting Istio support. 
+
+12. If the problem persists, capture the logs and escalate to OpenShift Serverless engineering team with a Knative ["must-gather"](https://github.com/openshift-knative/must-gather) dump.
```
Comment (skonto) [inline docs/sops/serverless-controller-knative-up.md:147]: Could you add the command here:

```oc adm must-gather --image=registry.redhat.io/openshift-serverless-1/svls-must-gather-rhel8 ```

That is the image I see people use. Not sure if there is a difference compared to quay.io/openshift-knative/must-gather.

Also it would be useful to add the SM one.

Comment (skonto) [reply docs/sops/serverless-controller-knative-up.md:147]: I think SM uses `registry.redhat.io/openshift-service-mesh/istio-must-gather-rhel8:2.4.0` but needs verification.

File: docs/sops/serverless-controller-knative-up.md [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+## Steps
+
+1. Check to see if all the pods are running in `openshift-serverless` namespace:
+
+```bash
+$ oc -n openshift-serverless get pods
+```
+
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
```
Comment (skonto) [inline docs/sops/serverless-controller-knative-up.md:60]: Maybe for events and particular pods this is more focused:
https://stackoverflow.com/questions/51931113/kubectl-get-events-only-for-a-pod

Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:60]: I followed existing SOPs and use them as template.

I think it is a good idea to keep this instruction more generic. We already give instructions in the sections above for collecting events and logs.

File: docs/sops/serverless-controller-knative-up.md (hunk 13 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:50]: I will make some adjustments to the text here.

Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md]: I mentioned ServiceMesh too now.

>  people should also check their setup there for issues
This is already written in step 11 below

File: docs/sops/serverless-controller-knative-up.md (hunk 14 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md]: Removed the mention to must gather.

Comment (skonto) [reply docs/sops/serverless-controller-knative-up.md]: They could also reach support or some contact?

File: docs/sops/serverless-controller-knative-up.md (hunk 15 above)
Comment (skonto) [reply docs/sops/serverless-controller-knative-up.md:60]: Ok sure.

File: docs/sops/serverless-controller-knative-up.md (hunk 14 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md]: Good feedback, thanks. Let me find some links to put in here :)

File: docs/sops/serverless-controller-knative-up.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
+# OpenShift Serverless Knative Serving Control Plane Readiness
```
Comment (skonto) [inline docs/sops/serverless-controller-knative-up.md:1]: Are you planning to add anything for Eventing too?

Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md:1]: Not right now. This SOP is needed for RHODS, but they don't use Eventing.

File: docs/sops/serverless-controller-knative-up.md (hunk 14 above)
Comment (aliok) [reply docs/sops/serverless-controller-knative-up.md]: @skonto added some links about how to reach out to SM team.
(I've asked in Istio team channel)

**Review by ReToCode — APPROVED** [review]
/lgtm
